etl/extractor/platform_impl.go

```diff
@@ -33,220 +30,160 @@ type PlatformExtractor struct {
-					request.RecordNumber = 0
-					receivedPulseNumber = request.PulseNumber.AsUint32()
-					log.Debugf("jump to pulse number: %v", receivedPulseNumber)
-					break
-				}
-
-				// save the last pulse for future requests
-				request.RecordNumber = resp.RecordNumber
-				request.PulseNumber = resp.Record.ID.Pulse()
-
-				receivedPulseNumber = request.PulseNumber.AsUint32()
-
-				// collect all records by PulseNumber
-				if receivedPulseNumber == lastPulseNumber {
-					jetDrops.Records = append(jetDrops.Records, resp)
-					continue
-				}
-
-				lastPulseNumber = receivedPulseNumber
-
-				// sending data to channel
-				m.mainJetDropsChan <- jetDrops
-				// zeroing variable which collecting jetDrops
-				jetDrops = new(types.PlatformJetDrops)
-				// don't forget to save the last data
-				jetDrops.Records = append(jetDrops.Records, resp)
-
-				if shouldReload && receivedPulseNumber > unsignedToPulseNumber {
-					// now we have received all needed data
-					return
-				}
+		pu, err = e.pulseExtractor.GetNextFinalizedPulse(ctx, int64(before.PulseNumber))
+		if err != nil { // network error ?
+			if strings.Contains(err.Error(), pulse.ErrNotFound.Error()) { // seems this pulse already last
+				time.Sleep(halfPulse)
+				continue
 			}
+			log.Errorf("retrievePulses(): before=%d err=%s", before, err)
+			pu = &before
+			time.Sleep(time.Second)
+			continue
 		}
-	}()
-}
-
-func (m *PlatformExtractor) getJetDropsContinuously(ctx context.Context) {
-	logger := belogger.FromContext(ctx)
-	var pulse uint32 = 0
-	var err error
-
-	// try to get current pulse with attempts
-	for i := 0; i < m.pulseExtractAttempts; i++ {
-		pulse, err = m.pulseExtractor.GetCurrentPulse(ctx)
-		if err != nil {
-			logger.Warnf("trying to get current pulse, attempt: %d", i)
-			time.Sleep(time.Duration(pulseDelta) * time.Second)
-		} else {
-			break
+		if pu.PulseNumber == before.PulseNumber { // no new pulse happens
+			time.Sleep(halfPulse)
+			continue
 		}
-	}
 
-	// fatal and exit if could not get current pulse
-	if pulse == 0 || err != nil {
-		logger.Fatalf("could not get current pulse number after %d attempts", m.pulseExtractAttempts)
-	}
-
-	logger.Infof("current pulse number: %d", pulse)
-	m.request.PulseNumber = insolar.PulseNumber(pulse)
-	m.request.RecordNumber = 0
-	m.getJetDrops(ctx, m.request, 0, math.MaxUint32, false)
-}
+		log.Debug("retrievePulses(): Done")
 
-func logGRPCError(ctx context.Context, err error) {
-	log := belogger.FromContext(ctx)
-	if err != nil {
-		log.Debug("Data request failed: ", err)
-		log.Error(errors.Wrapf(err, "failed to get gRPC stream from exporter.Export method").Error())
-	}
-}
+		go e.retrieveRecords(ctx, pu)
 
-func logIfErrorReceived(ctx context.Context, err error, request interface{}) {
-	if err != nil {
-		belogger.FromContext(ctx).Debug("received error value from records gRPC stream %v", request)
-	}
-}
+		if until <= 0 { // we are going on the edge of history
+			time.Sleep(halfPulse)
```

Comment:
if we get missed pulses, should we sleep here for so long? Or its for not loading heavy too much?

Reply:
if until <= 0

Reply:
Yes, in fact we need to control it "max parallel streams" way, but not with pauses.